compiler/qsc_data_structures/src/line_column.rs

```diff
@@ -0,0 +1,192 @@
+    /// as byte offsets, and [`Encoding::Utf16`] will use utf-16 code unit (word) offsets.
+    ///
+    /// If the position is past the end of the string, returns the offset of
+    /// the end of the string (e.g. for "hello" returns 5).
+    #[must_use]
+    pub fn to_utf8_byte_offset(&self, encoding: Encoding, contents: &str) -> u32 {
+        let mut column: u32 = 0;
+        let mut line: u32 = 0;
+
+        for (byte_offset, c) in contents.char_indices() {
+            if c == '\n' {
+                line += 1;
+                column = 0;
+            } else {
+                column += num_code_units(encoding, c);
+            }
+
+            if line > self.line || (line == self.line && column > self.column) {
+                // We moved past the target line+column
+                return u32(byte_offset);
+            }
+        }
+
+        // return eof if we move past the end of the string
+        u32(contents.len())
+    }
+}
+
+/// Same as a [`Span`], but represented with [`Position`]s instead of offsets.
+#[derive(Debug, PartialEq, Clone, Copy, Eq, Hash)]
+pub struct Range {
+    pub start: Position,
+    pub end: Position,
+}
+
+impl Range {
+    /// For a given string and a span expressed in utf-8 byte offsets,
+    /// returns a [`Range`] that corresponds to that span.
+    ///
+    /// Any offsets that go past the end of the string are mapped to the end of the string.
+    /// (e.g. for "hello", the span (0-10) would map to (0,0-0,5)).
+    ///
+    /// This function does not validate span range. If `span.hi > span.lo`, the
+    /// end offset is simply mapped to the end of the string.
+    /// (e.g. for "hello", the invalid span (3-0) would map to (0,3-0,5)).
+    ///
+    /// This function also does not validate whether the passed in offset
+    /// is a valid utf8 char boundary. If an invalid offset is passed in,
+    /// the next char boundary will be returned.
+    #[must_use]
+    pub fn from_span(encoding: Encoding, contents: &str, span: &Span) -> Self {
+        let [start, end] = positions_from_utf8_byte_offsets(encoding, contents, [span.lo, span.hi]);
+        Self { start, end }
+    }
+
+    #[must_use]
+    pub fn empty(&self) -> bool {
+        self.start == self.end
+    }
+}
+
+/// For a given string and array of utf-8 byte offsets, returns the [`Position`]s
+/// corresponding to the byte offsets.
+///
+/// Since this function iterates over the string once, the array *MUST* be in sorted order.
+///
+/// This function takes a (const sized) array to avoid allocations.
+fn positions_from_utf8_byte_offsets<const N: usize>(
+    encoding: Encoding,
+    contents: &str,
+    sorted_utf8_byte_offsets: [u32; N],
+) -> [Position; N] {
+    // The below example contains characters that are encoded
+    // with more utf-8 code units (bytes) than utf-16 code units, to demonstrate
+    // how code unit offset will differ depending on the chosen encoding.
```

Comment:
It's probably also worth calling and out including examples of where there are more utf-16 code units than characters, such as those above the BMP from U+010000 to U+10FFFF (where one code point is encoded as a UTF-16 'surrogate pair' of code units between 0xD800 and 0xDFFF - see https://en.wikipedia.org/wiki/UTF-16 )

Reply:
(I guess the first char below shows that, but worth a note in the comments maybe)